### PR TITLE
add end to end stack depth test

### PR DESCRIPTION
**Pull request by TyOverby**
Related bug: #16669 

(Test only change)

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public class EndToEndTests: EmitMetadataTestBase
+    {
+        [WorkItem(16669, "https://github.com/dotnet/roslyn/issues/16669")]
+        [Fact]
+        public void OverflowOnFluentCall()
+        {
+#if DEBUG
+            int numberFluentCalls = 290;
+#else
+            int numberFluentCalls = 590;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: @jcouv test said we could hit 630.  Is 600 not even working here?

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+using System.Threading.Tasks;
+using Xunit;
+using Microsoft.CodeAnalysis.Test.Utilities;
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Emit
+{
+    public class EndToEndTests: EmitMetadataTestBase
+    {
+        [WorkItem(16669, "https://github.com/dotnet/roslyn/issues/16669")]
+        [Fact]
+        public void OverflowOnFluentCall()
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: Can we add a comment indicating we want to ensure that we don't regress the level of fluent calls allowed here.  Make it less of a surprise when developers hit this.

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                return builder.ToString();
+            }
+            var source = MakeCode();
+
+            var thread = new System.Threading.Thread(() =>
+            {
+                var compilation = CreateCompilationWithMscorlib(source);
+                compilation.GetDiagnostics();
+                compilation.EmitToArray();
+            }, 0);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: Reading the [doc on maxThreadSize](https://msdn.microsoft.com/en-us/library/5cykbwz4(v=vs.110).aspx#Anchor_2), it seems that 0 means falling back to minimum stack size, which is a page. Isn't that going to be system-dependent (ie error-prone)?

Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: Good catch.  We should be explicit here.  

Also should use a named argument.

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs (hunk 1 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: anything over 590 breaks for me.

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs (hunk 3 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: I would expect max-stack-size to be pretty error prone, yeah.

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+}");
+                return builder.ToString();
+            }
+            var source = MakeCode();
+
+            var thread = new System.Threading.Thread(() =>
+            {
+                var compilation = CreateCompilationWithMscorlib(source);
+                compilation.GetDiagnostics();
+                compilation.EmitToArray();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: `compilation.VerifyDiagnostics();`

Comment (TyOverby) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: Done

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs (hunk 2 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: Done

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs (hunk 3 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: > Good catch. We should be explicit here.

Sure, but if we go over the maxThreadSize, it falls back without any notification.  So we'd still 

> maxStackSize is ignored if it is greater than the default stack size. No exception is thrown. [msdn](https://msdn.microsoft.com/en-us/library/5cykbwz4(v=vs.110).aspx)

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs (hunk 1 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: @jaredpar We figured out the discrepancy (in my setup, the test itself was built in Debug, while the libraries it used were Retail). With the same setup, Ty gets the same number.

**Comment by TyOverby** [comment]
Well.  Looks like the Jenkins machines fail this test.  This is going to be miserable for devs if things start breaking.

@jaredpar Another option for us is to add a "stack depth" test on the perf system.  Let it track the number of fluent calls that it takes to break the compiler.  This way we'll have graphs :P

**Comment by TyOverby** [comment]
@jaredpar Still fails for some reason on the release build.  Should I just keep committing until I find out what the limit for the Azure machines are?

**Comment by jaredpar** [comment]
> Should I just keep committing until I find out what the limit for the Azure machines are?

Did you verify that changing parallel moved the stack overflow to the main thread?  That would be the first step.  If it's on the main thread then it should be consistent across machines.  If it's not then we need to look at a different way to get it onto a thread that we control.

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            }
+            var source = MakeCode();
+
+            var thread = new System.Threading.Thread(() =>
+            {
+                var options = new CSharpCompilationOptions(outputKind: OutputKind.DynamicallyLinkedLibrary, concurrentBuild: false);
+                var compilation = CreateCompilationWithMscorlib(source, options: options);
+                compilation.VerifyDiagnostics();
+                compilation.EmitToArray();
+            }, 0);
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: From https://msdn.microsoft.com/en-us/library/5cykbwz4(v=vs.110).aspx

> Avoid using this constructor overload. The default stack size used by the [`Thread(ThreadStart)`](https://msdn.microsoft.com/en-us/library/xx3ezzs2(v=vs.110).aspx) constructor overload is the recommended stack size for threads. If a thread has memory problems, the most likely cause is programming error, such as infinite recursion.

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            thread.Start();
+            thread.Join();
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: Does the stack overflow in that thread cause this test to fail? What is the failure mode?

**Comment by TyOverby** [comment]
@jaredpar Wait, why do we want this on the main thread?  I added an explicit Thread.Start so that there wouldn't be anything else on the stack.

**Comment by jaredpar** [comment]
@TyOverby 

By main thread I mean the thread started by `Thread.Start`.  

The point of starting a new thread here is so we can define a consistent environment for validating or minimum requirements on stack depth.  Using `Task.Run` or similar APIs isn't good enough because the depth could vary across implementations / operating systems. 

This only works though if the binding code, the code which is presently overflowing, is run on this thread.  That's not happening in the current iteration of the test because the compiler is essentially spawning that code off via the thread pool.  We need to change this test so the binding code runs on a thread we control.

**Comment by TyOverby** [comment]
I've confirmed that the exception is thrown from the same thread that we've spawned.

**Comment by jaredpar** [comment]
Are you running locally with 32 bit? Is that the reason for the diff?  The 32 bit suites are passing.

**Comment by TyOverby** [comment]
> Are you running locally with 32 bit? Is that the reason for the diff? The 32 bit suites are passing.

Argh, that was it.  I don't know why I assumed that we'd default to 64-bit in our visual studio start-up projects.

**Comment by jaredpar** [comment]
You know what happens when you assume?  Visual Studio picks the other CPU affinity that's what happens! 😄

**Comment by jaredpar** [comment]
@TyOverby how is this change going?

**Comment by TyOverby** [comment]
@jaredpar @jcouv @cston: tests are passing!

**Comment by jcouv** [comment]
Nice. 
I think it is better to wait for the fix to go in (https://github.com/dotnet/roslyn/pull/16795), then we can check in the test with the higher limit(s).

**Comment by TyOverby** [comment]
@jcouv: sounds good.  I'll wait on that.

**Comment by TyOverby** [comment]
Actually @jcouv, do you want me to make a PR against your PR, adding this test?

**Comment by jcouv** [comment]
No need. We'll know within an hour whether the fix goes into RTM or not.

**Comment by jcouv** [comment]
FYI fix was merged into master

File: src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs (hunk 6 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Test/Emit/Emit/EndToEndTests.cs]: Yeah, the exception causes the test to fail.

**Comment by TyOverby** [comment]
Cool.  Merging now.

**Comment by jaredpar** [comment]
Why didn't we expand the values to take advantage of the new depth?

**Comment by TyOverby** [comment]
@jaredpar You made it sound like this test was required for ask-mode.  I wanted to get it in as soon as the other PR landed so we won't worry about regressions.  I'll make a followup PR right now.

**Pull request merged**